### ISSUE TITLE
`clippy::map_flatten` multiline suggestion is missing closing tokens

### ISSUE THREAD

**Issue by CBenoit** [rust-lang/rust-clippy#8506]
### Summary

Suggestion is incorrect Rust code.

### Reproducer

I tried this code:

```rust
Some(5)
    .map(|some_long_variable_name| {
        if some_long_variable_name <= 5 {
            Some(some_long_variable_name)
        } else {
            None
        }
    })
    .flatten();
```

I expected to see this happen:
```
warning: called `map(..).flatten()` on an `Option`
  --> picky-signtool/src/main.rs:14:12
   |
14 |       Some(5)
   |  ____________^
15 | |         .map(|some_long_variable_name| {
16 | |             if some_long_variable_name <= 5 {
17 | |                 Some(some_long_variable_name)
...  |
21 | |         })
22 | |         .flatten();
   | |__________________^
   |
   = note: `#[warn(clippy::map_flatten)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_flatten
help: try using `and_then` instead
   |
14 ~     Some(5).and_then(|some_long_variable_name| {
15 +             if some_long_variable_name <= 5 {
16 +                 Some(some_long_variable_name)
17 +             } else {
18 +                 None
19 +             }
20 +     });
 ...
```

Instead, this happened:

```
warning: called `map(..).flatten()` on an `Option`
  --> picky-signtool/src/main.rs:14:12
   |
14 |       Some(5)
   |  ____________^
15 | |         .map(|some_long_variable_name| {
16 | |             if some_long_variable_name <= 5 {
17 | |                 Some(some_long_variable_name)
...  |
21 | |         })
22 | |         .flatten();
   | |__________________^
   |
   = note: `#[warn(clippy::map_flatten)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_flatten
help: try using `and_then` instead
   |
14 ~     Some(5).and_then(|some_long_variable_name| {
15 +             if some_long_variable_name <= 5 {
16 +                 Some(some_long_variable_name)
17 +             } else {
18 +                 None
19 +             }
 ...
```

### Version

```text
rustc 1.59.0 (9d1b2106e 2022-02-23)
binary: rustc
commit-hash: 9d1b2106e23b1abd32fce1f17267604a5102f57a
commit-date: 2022-02-23
host: x86_64-unknown-linux-gnu
release: 1.59.0
LLVM version: 13.0.0
```

**Comment by J-ZhengLi** [comment]
@xFrednet Sorry for the pin, but I'm confused about how should I implement this. Since this seems to be a design flaw of [`rustc_errors::emitter`](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_errors/emitter.rs.html#552), until someone has find better solution, should I try to hide the suggestion code while displaying it as help message using `span_lint_and_then`?

**Comment by xFrednet** [comment]
Hey @J-ZhengLi, no worries you're always welcome to ping me :upside_down_face: This is a "design flaw" at one pint it might be good to add a new display option to rustc to say that the edges of the marked span are important. For now we can check if the suggestion is larger than `MAX_SUGGESTION_HIGHLIGHT_LINES` and if so adjust the suggestion to only show the important snippets. In this case you can create a suggestion like this:

```text
warning: called `map(..).flatten()` on an `Option`
  --> picky-signtool/src/main.rs:14:12
   |
14 |       Some(5)
   |  ____________^
15 | |         .map(|some_long_variable_name| {
16 | |             if some_long_variable_name <= 5 {
17 | |                 Some(some_long_variable_name)
...  |
21 | |         })
22 | |         .flatten();
   | |__________________^
   |
   = note: `#[warn(clippy::map_flatten)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_flatten
help: try using `and_then` instead
   |
14 ~     Some(5).and_then(|some_long_variable_name| {
15 +             if some_long_variable_name <= 5 {
16 +                 Some(some_long_variable_name)
   |
help: and remove the final `.flatten()`
   |
18 +                 None
19 +             }
20 ~         });
   |
```

You can also create a helper method that can be shared between lints if you want to be fancy. The proper way would be to have this in rustc, but I just thought of that. The suggested method formatting will be a good solution in the meantime :upside_down_face:
